src/en/reference-hook-tools.md

```diff
@@ -0,0 +1,564 @@
+      * only the leader can write to the bucket
+      * only minions are informed of changes to the bucket
+      * changes are propagated instantly, bypassing the sandbox
+
+The instant propagation may be surprising, but it exists to satisfy the use case
+where shared data can be chosen by the leader at the very beginning of (say)
+the install hook. By propagating it instantly, any running minions can make
+use of the data and progress immediately, without having to wait for the
+leader to finish its hook.
+
+It also means that you can guarantee that a successful `leader-set` call has
+been reflected in the database, and that all minions will converge towards
+seeing that value, even if an unexpected error takes down the current hook.
+
+For both these reasons it is strongly recommended that leader settings are 
+always written as a self-consistent group (`leader-set foo=bar baz=qux ping=pong`,
+rather than `leader-set foo=bar; leader-set baz=qux` etc, to avoid situations
+where minions may end up seeing a sandbox in which only `foo` is set to the 
+"correct" value).
+
+### leader-get
+
+`leader-get` acts much like relation-get, in that it lets you read string
+values by key (and expose them in helpful formats), but it reads only from the 
+single leader-settings bucket
+
+As with `realtion-get`, it presents a sandboxed view of leader-settings data. 
+This is necessary, as it is for relation data, because a hook context needs
+to present *consistent* data; but it means that there's a small extra burden
+on users of `leader-set`.
+
+
+### open-port
+
+`open-port` marks a port or range of ports on the local system as appropriate to
+open, if and when the service is exposed to the outside world. It accepts a
+single port or range of ports with an optional protocol, which may be `udp` or
+`tcp`, where `tcp` is the default.
+
+Examples:
+
+Open 80/tcp if and when the service is exposed:
+
+```no-highlight
+open-port 80
+```
+
+Open 1234/udp if and when the service is exposed:
+
+```no-highlight
+open-port 1234/udp
+```
+
+Open the range 8000 to 8080:
+
+```no-highlight
+open 8000-8080/tcp
+```
+
+`open-port` will not have any effect if the service is not exposed, and may have
+a somewhat delayed effect even if it is. This operation is transactional, so 
+changes will certainly not be made unless the hook exits successfully.
+
+Juju also tracks ports opened across the machine and will not allow conflicts - 
+if another charm has already opened the port 
+(**or one or more ports in a range**) you have specified,
+your request will be ignored. 
+ 
+This command accepts and ignores `--format` for
+compatibility purposes, but it doesn't produce any output.
+
+### opened-ports
+
+The opened-ports hook tool lists all the ports currently opened 
+**by the running charm**. It does not, at the moment, include ports which may
+be opened by other charms co-hosted on the same machine
+[lp#1427770](https://bugs.launchpad.net/juju-core/+bug/1427770). 
```

Comment:
Comment: We should probably standardize on how we want to link to bugs. Under src we have these formats present:

[LP #1450092]
[LP bug #1347020]
[lp#1427770]